src/server-logger.ts

```diff
@@ -0,0 +1,257 @@
+      const newTrace = this.truncateLog(logLevel, logMessage);
+      this.logger.warn('log message too large to send to server. truncating log details and/or message', {
+        originalTrace: (trace as any),
+        truncatedTrace: (newTrace as any)
+      }, true);
+
+      /* newTrace will be `null` if the truncated trace was still too big */
+      if (newTrace === null) {
+        this.logger.error('truncated message is still too large to send to server. not sending message', {
+          originalTrace: (trace as any)
+        }, true);
+        return;
+      }
+
+      /* set the trace to our new trunctated trace item */
+      traceMessageSize = calculateLogMessageSize(newTrace);
+      trace = newTrace;
+    }
+
+    /* use the last item in the buffer if it exists, otherwise start with a blank buffer item */
+    let bufferItem: ILogBufferItem;
+    if (this.logBuffer.length) {
+      bufferItem = this.logBuffer[this.logBuffer.length - 1];
+    } else {
+      bufferItem = {
+        size: 0,
+        traces: []
+      };
+    }
+
+    /* if pushing our trace onto the buffer item will be too large, we need a new buffer item */
+    const exceedsMaxLogSize = bufferItem.size + traceMessageSize > MAX_LOG_SIZE;
+    if (exceedsMaxLogSize) {
+      // console.log('%c `exceedsMaxLogSize` was `true`', 'color: #32a852', { logBuffer: [...this.logBuffer], bufferItem, incomingTrace: trace, incomingTraceSize: traceMessageSize }); // TODO: delete me
+      this.logBuffer.push({
+        size: traceMessageSize,
+        traces: [trace]
+      });
+      /* since we pushed a new item, we need to send immediately */
+      // console.log('%c calling sendLogsToServer(true)', 'color: #32a852', { logBuffer: [...this.logBuffer] }); // TODO: delete me
+      this.sendLogsToServer(true);
+      return;
+    }
+
+    /* else just push onto the buffer */
+    bufferItem.size += traceMessageSize;
+    bufferItem.traces.push(trace);
+
+    /* if we don't have anything in the buffer, that means we have to push this new item */
+    if (!this.logBuffer.length) {
+      // console.log('%c `this.logBuffer` was empty. pushing new buffer item', 'color: #32a852', { logBuffer: [...this.logBuffer], bufferItem }); // TODO: delete me
+      this.logBuffer.push(bufferItem);
+    }
+
+    /* this will setup the debounce timer (if it is not already running) */
+    // console.log('%c calling sendLogsToServer()', 'color: #32a852', { logBuffer: [...this.logBuffer] }); // TODO: delete me
+    this.sendLogsToServer();
+  }
+
+  private async sendLogsToServer (immediate: boolean = false): Promise<any> {
+    if (!this.logBuffer.length) {
+      /* clear timer */
+      clearTimeout(this.debounceTimer);
+      this.debounceTimer = null;
+      // console.log('%c buffer empty, not sending http request', 'color: #32a852'); // TODO: delete me
+      return;
+    }
+
+    /* if we have a pending request, then don't send another */
+    if (this.pendingHttpRequest) {
```

Comment:
would it be reasonable to try and send if `immediate === true` and we already have a pendingHttpRequest?

Reply:
I'm open to the idea. It would take a little more logic. Curious what the benefit would be from having multiple requests sending?

Reply:
I see what you are saying now. I was trying to have all the retry logic here in the `server-logger` but now that I understand the use of the `log-uploader`, I am going to move the logic into there. I won't need to track any `pendingHttpRequest` here. And if we push multiple requests through, the uploader will handle the retry and pending requests logic.